src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -91,55 +83,75 @@ pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context
-    let idx_in_bits = &idx.view_bits::<Lsb0>()[..num_multi_bits];
-
-    try_join_all(
-        (0..num_records)
-            .zip(repeat(ctx_across_bits))
-            .map(|(rec, ctx)| async move {
-                let share_of_one = ctx.share_of_one();
-
-                let mult_input = zip(multi_bit_input, idx_in_bits)
-                    .map(|(single_bit_input, bit)| {
-                        if *bit {
-                            single_bit_input[rec].clone()
-                        } else {
-                            -single_bit_input[rec].clone() + &share_of_one
-                        }
-                    })
-                    .collect::<Vec<_>>();
-
-                multiply_all_shares(
-                    ctx,
-                    RecordId::from(idx * num_records + rec),
-                    mult_input.as_slice(),
-                )
-                .await
-            }),
-    )
-    .await
+    let record_id = RecordId::from(record_idx);
+    let num_bits = input.len();
+    let mut precomputed_combinations = Vec::with_capacity(1 << num_bits);
+    precomputed_combinations.push(ctx.share_of_one());
+    #[allow(clippy::needless_range_loop)]
+    for bit_idx in 0..num_bits {
+        let bit = &input[bit_idx][record_idx];
+        let step = 1 << bit_idx;
+        let num_children_to_add = precomputed_combinations.len();
+        precomputed_combinations.push(bit.clone());
+        if bit_idx == 0 {
+            continue;
+        }
+        let mut multiplication_futures = Vec::with_capacity(num_children_to_add - 1);
+        #[allow(clippy::needless_range_loop)]
+        for j in 1..num_children_to_add {
+            let child_idx = j + step;
+            multiplication_futures.push(ctx.narrow(&BitOpStep::from(child_idx)).multiply(
+                record_id,
+                &precomputed_combinations[j],
+                bit,
+            ));
+        }
+        let mut multiplication_results = try_join_all(multiplication_futures).await?;
+        precomputed_combinations.append(&mut multiplication_results);
+    }
+
+    let mut equality_checks = Vec::with_capacity(1 << num_bits);
+    for i in 0..(1 << num_bits) {
+        equality_checks.push(check_equality_to(
+            i,
+            num_bits,
+            precomputed_combinations.as_slice(),
+            0,
+            0,
+        ));
+    }
+    Ok(equality_checks)
+}
+
+fn check_equality_to<F: Field, S: SecretSharing<F>>(
+    value: u32,
+    num_bits: usize,
+    tree: &[S],
+    bit_idx: usize,
+    idx: usize,
+) -> S {
+    if bit_idx == num_bits {
+        return tree[idx].clone();
+    }
+    let bit = (value >> bit_idx) & 1;
+    let step = 1 << bit_idx;
+    let next_bit_idx = bit_idx + 1;
+    let times_bit = check_equality_to(value, num_bits, tree, next_bit_idx, idx + step);
+    if bit == 0 {
+        let times_one = check_equality_to(value, num_bits, tree, next_bit_idx, idx);
+        return times_one - &times_bit;
+    }
+    times_bit
```

Comment:
```suggestion
) fn check_equality_to<F: Field, S: SecretSharing<F>>(
    value: u32,
    num_bits: usize,
    precomputed_combinations: &[S],
    bit_idx: usize,
    tree_idx: usize,
) -> S {
    if bit_idx == num_bits {
        return precomputed_combinations[tree_idx].clone();
    }
    // 1<<bit_idx + tree_idx gives the right subtree equivalent for this bit (i.e. for 1-b it finds location of b)
    let times_bit = check_equality_to(value, num_bits, precomputed_combinations, bit_idx + 1, tree_idx + 1 << bit_idx);
    let bit = (value >> bit_idx) & 1;
    if bit == 0 {
        let times_one = check_equality_to(value, num_bits, precomputed_combinations, bit_idx + 1, tree_idx);
        return times_one - &times_bit;
    }
    times_bit
}
```
This isnt the best, but atleast helps know what each branch is trying to do. Some diagram would definitely be helpful here